### PR TITLE
Plfm 4890 e

### PR DESCRIPTION
**Pull request by john-hill**

### CODE REVIEW

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/utils/TableModelUtils.java [hunk 1]
```diff
@@ -78,7 +81,7 @@ public String apply(Long input) {
 	public static final String EXCEEDS_MAX_SIZE_TEMPLATE = "Request exceeds the maximum number of bytes per request.  Maximum : %1$s bytes";
 
 	private static final String INVALID_VALUE_TEMPLATE = "Value at [%1$s,%2$s] was not a valid %3$s. %4$s";
-	private static final String TABLE_SEMAPHORE_KEY_TEMPLATE = "TALBE-LOCK-%1$d";
+	private static final String TABLE_SEMAPHORE_KEY_TEMPLATE = "TALBE-LOCK-%1$d%2$s";
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/utils/TableModelUtils.java:84]: remove

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/utils/TableModelUtils.java [hunk 2]
```diff
@@ -64,6 +65,8 @@
  */
 public class TableModelUtils {
 
+	private static final String TALBE_LOCK_PREFIX = "TALBE-LOCK-";
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/utils/TableModelUtils.java:68]: table

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableEntityManagerImpl.java [hunk 3]
```diff
@@ -837,22 +843,36 @@ public Long getChangeNumber() {
 		public TableChangeType getChangeType() {
 			return wrapped.getChangeType();
 		}
+
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableEntityManagerImpl.java:846]: use array list for line 810 with size of input list.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableIndexManager.java [hunk 4]
```diff
@@ -187,6 +191,19 @@ public Long populateViewFromEntityReplication(Long tableId, ProgressCallback cal
...
+	 * including the provided last change number.
+	 * @param progressCallback 
+	 * 
+	 * @param tableId
+	 * @param iterator
+	 * @param lastChangeNumber
+	 * @throws RecoverableMessageException Will RecoverableMessageException if the index cannot be built at this time.
+	 */
+	public void buildIndexToChangeNumber(ProgressCallback progressCallback, IdAndVersion idAndVersion, Iterator<TableChangeMetaData> iterator,
+			Optional<Long> lastChangeNumber) throws RecoverableMessageException;
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableIndexManager.java:206]: lastChangeNumber is not actually optional.   Force the caller to not call this method if the value is they have is null.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableIndexManagerImpl.java [hunk 5]
```diff
@@ -354,5 +369,126 @@ ColumnModelPage getPossibleAnnotationDefinitionsForContainerIds(
 		results.setResults(columns);
 		return results;
 	}
+	
+	@Override
+	public void buildIndexToChangeNumber(final ProgressCallback progressCallback, final IdAndVersion idAndVersion,
+			final Iterator<TableChangeMetaData> iterator, final Optional<Long> lastChangeNumber) throws RecoverableMessageException {
+		// Only proceed if work is needed.
+		if (!tableManagerSupport.isIndexWorkRequired(idAndVersion)) {
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableIndexManagerImpl.java:377]: This needs to the last change number to correctly answer this question.

Also rename lastChangeNumber -> targetChangeNumber

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableIndexManagerImpl.java [hunk 6]
```diff
@@ -354,5 +369,126 @@ ColumnModelPage getPossibleAnnotationDefinitionsForContainerIds(
 		results.setResults(columns);
 		return results;
 	}
+	
+	@Override
+	public void buildIndexToChangeNumber(final ProgressCallback progressCallback, final IdAndVersion idAndVersion,
+			final Iterator<TableChangeMetaData> iterator, final Optional<Long> lastChangeNumber) throws RecoverableMessageException {
+		// Only proceed if work is needed.
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableIndexManagerImpl.java:376]: will throw illegal arg if last change number is null.

File: services/workers/src/main/java/org/sagebionetworks/table/worker/TableIndexWorker.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+				throw new RecoverableMessageException();
+			}
+			if (ChangeType.DELETE.equals(message.getChangeType())) {
+				// Delete the table in the index
+				tableEntityManager.deleteTableIfDoesNotExist(tableId);
+				indexManager.deleteTableIndex(idAndVersion);
+				return;
+			} else {
+				// Build the table's index.
+				Optional<Long> lastChangeNumber = tableEntityManager.getLastTableChangeNumber(tableId);
```
Comment (john-hill) [inline services/workers/src/main/java/org/sagebionetworks/table/worker/TableIndexWorker.java:51]: if null then do nothing.

File: services/workers/src/main/java/org/sagebionetworks/table/worker/TableIndexWorker.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import org.sagebionetworks.repo.manager.table.change.TableChangeMetaData;
+import org.sagebionetworks.repo.model.ObjectType;
+import org.sagebionetworks.repo.model.entity.IdAndVersion;
+import org.sagebionetworks.repo.model.message.ChangeMessage;
+import org.sagebionetworks.repo.model.message.ChangeType;
+import org.sagebionetworks.workers.util.aws.message.RecoverableMessageException;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class TableIndexWorker implements ChangeMessageDrivenRunner {
+
```
Comment (john-hill) [inline services/workers/src/main/java/org/sagebionetworks/table/worker/TableIndexWorker.java:22]: need a test for this worker

**Review by zimingd — APPROVED** [review]

**Pull request merged**